sky/provision/docker_utils.py

```diff
@@ -0,0 +1,347 @@
+
+    @classmethod
+    def from_env_vars(cls, d: Dict[str, str]) -> 'DockerLoginConfig':
+        return cls(
+            username=d[constants.DOCKER_USERNAME_ENV_VAR].strip(),
+            password=d[constants.DOCKER_PASSWORD_ENV_VAR].strip(),
+            server=d[constants.DOCKER_SERVER_ENV_VAR].strip(),
+        )
+
+
+# Copied from ray.autoscaler._private.ray_constants
+# The default maximum number of bytes to allocate to the object store unless
+# overridden by the user.
+DEFAULT_OBJECT_STORE_MAX_MEMORY_BYTES = 200 * 10**9
+# The default proportion of available memory allocated to the object store
+DEFAULT_OBJECT_STORE_MEMORY_PROPORTION = 0.3
+
+
+def _check_helper(cname, template, docker_cmd):
+    return ' '.join([
+        docker_cmd, 'inspect', '-f', '"{{' + template + '}}"', cname, '||',
+        'true'
+    ])
+
+
+def check_docker_running_cmd(cname, docker_cmd):
+    return _check_helper(cname, '.State.Running', docker_cmd)
+
+
+def check_bind_mounts_cmd(cname, docker_cmd):
+    return _check_helper(cname, 'json .Mounts', docker_cmd)
+
+
+def check_docker_image(cname, docker_cmd):
+    return _check_helper(cname, '.Config.Image', docker_cmd)
+
+
+def docker_start_cmds(
+    image,
+    container_name,
+    user_options,
+    docker_cmd,
+):
+    """Generating docker start command without --rm.
+
+    The code is borrowed from `ray.autoscaler._private.docker`. The only
+    difference is that we don't use `--rm` flag here.
+    """
+
+    # for click, used in ray cli
+    env_vars = {'LC_ALL': 'C.UTF-8', 'LANG': 'C.UTF-8'}
+    env_flags = ' '.join(
+        ['-e {name}={val}'.format(name=k, val=v) for k, v in env_vars.items()])
+
+    user_options_str = ' '.join(user_options)
+    docker_run = [
+        docker_cmd,
+        'run',
+        # SkyPilot: Remove --rm flag to keep the container after `ray stop`
+        # is executed.
+        '--name {}'.format(container_name),
+        '-d',
+        '-it',
+        env_flags,
+        user_options_str,
+        '--net=host',
+        image,
+        'bash',
+    ]
+    return ' '.join(docker_run)
+
+
+def _with_interactive(cmd):
+    force_interactive = (
+        f'source ~/.bashrc; '
+        f'export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && ({cmd})')
+    return ['bash', '--login', '-c', '-i', shlex.quote(force_interactive)]
+
+
+# SkyPilot: New class to initialize docker containers on a remote node.
+# Adopted from ray.autoscaler._private.command_runner.DockerCommandRunner.
+class DockerInitializer:
+    """Initializer for docker containers on a remote node."""
+
+    def __init__(self, docker_config: dict,
+                 runner: 'command_runner.SSHCommandRunner', log_path: str):
+        self.docker_config = docker_config
+        self.container_name = docker_config['container_name']
+        self.runner = runner
+        self.home_dir = None
+        self.initialized = False
+        use_podman = docker_config.get('use_podman', False)
```

Comment:
I'm thinking shall we deprecate this podman option..? At least only docker is tested for now...
But considering it won't affect correctness, I think it is ok to keep it 🫡